[CSS Masking] SVG masks are not working as 'mask-image'

shallawa opened the pull request:
#### cbbe5238d695fcf62431ab47772f5f4fe25a0559
<pre>
[CSS Masking] SVG masks are not working as &apos;mask-image&apos;
<a href="https://bugs.webkit.org/show_bug.cgi?id=260732">https://bugs.webkit.org/show_bug.cgi?id=260732</a>
rdar://114465545

Reviewed by Simon Fraser.

Implement CSS masking when it references an SVG masker. There seven cases for
mask-image values that need to be considered:

1. Bitmap image
2. Bitmap image fragment
3. External SVG image
4. External SVG image fragment
5. External SVG image mask
6. Inline data URL image
7. Inline SVG image mask

StyleCachedImage will be adjusted to take into account the seventh case. Functions
like StyleCachedImage::isLoaded() and StyleCachedImage::canRender() will have to
handle the new case where StyleCachedImage represents an inline referenced SVG
masker. So isLoaded() needs to a take a RenderElement to be able to resolve the
imageURL.

A new super-class of GeneratedImage named SVGResourceImage will be introduced to
handle drawing the SVG resource given its renderer. SVGResourceImage will handle
the cases of external and inline SVG resources.

We can&apos;t know whether the ID associated with an external image is a fragment ID or
a mask ID in an SVGImage till the CachedImage is loaded. StyleCachedImage::image()
is called only after the image is completely loaded. So it can decide to create
an SVGResourceImage or to let the CachedImage creates its Image.

* LayoutTests/TestExpectations:
Some of the WPT tests have been marked as ImageOnlyFailure in this patch because
the expected results are now working as expected but the tests are not.
Unfortunately the tests and the expected results are both using mask-image. So
without this change they match because they both draw nothing for the masked
elements. With this change, the tests are failing  because the mask geometry
is set to work with content-box only. This can be addressed in a separate patch.

* LayoutTests/css3/masking/mask-reference-png-fragment-expected.html: Added.
* LayoutTests/css3/masking/mask-reference-png-fragment.html: Added.
* LayoutTests/css3/masking/resources/red-100x100.png: Added.
* LayoutTests/http/tests/css/css-masking/mask-external-svg-fragment-expected.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-external-svg-fragment.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-external-svg-image-expected.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-external-svg-image.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-external-svg-mask-expected.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-external-svg-mask.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-inline-svg-image-expected.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-inline-svg-image.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-inline-svg-mask-expected.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-inline-svg-mask.html: Added.
* LayoutTests/http/tests/css/css-masking/resources/heart.svg: Added.
* LayoutTests/http/tests/css/css-masking/resources/star-heart-fragments.svg: Added.
* LayoutTests/http/tests/css/css-masking/resources/star-heart-masks.svg: Added.
* LayoutTests/http/tests/css/css-masking/resources/star.svg: Added.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/mask-image/mask-opacity-1d.html:
This tests is now drawing the masked element but with noise around the borders.

* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSImageValue.cpp:
* Source/WebCore/platform/graphics/Image.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::isSVGResourceImage const):
* Source/WebCore/rendering/BorderPainter.cpp:
(WebCore::BorderPainter::paintBorder):
(WebCore::BorderPainter::paintNinePieceImage):
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::referenceFilterElement):
* Source/WebCore/rendering/InlineBoxPainter.cpp:
(WebCore::InlineBoxPainter::paintMask):
(WebCore::InlineBoxPainter::paintDecorations):
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
(WebCore::ReferencedSVGResources::referencedRenderResource):
* Source/WebCore/rendering/ReferencedSVGResources.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::paintMaskImages):
(WebCore::RenderBox::repaintLayerRectsForImage):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::borderImageIsLoadedAndCanBeRendered const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::setupClipPath):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::canDirectlyCompositeBackgroundBackgroundImage):
(WebCore::hasPaintedBoxDecorationsOrBackgroundImage):
(WebCore::supportsDirectlyCompositedBoxDecorations):
(WebCore::RenderLayerBacking::contentChanged):
* Source/WebCore/rendering/style/FillLayer.cpp:
(WebCore::FillLayer::imagesAreLoaded const):
* Source/WebCore/rendering/style/FillLayer.h:
* Source/WebCore/rendering/style/NinePieceImage.cpp:
(WebCore::NinePieceImage::paint const):
* Source/WebCore/rendering/style/StyleCachedImage.cpp:
(WebCore::StyleCachedImage::renderSVGResource const):
(WebCore::StyleCachedImage::canRender const):
(WebCore::StyleCachedImage::isLoaded const):
(WebCore::StyleCachedImage::imageSize const):
(WebCore::StyleCachedImage::computeIntrinsicDimensions):
(WebCore::StyleCachedImage::setContainerContextForRenderer):
(WebCore::StyleCachedImage::image const):
* Source/WebCore/rendering/style/StyleCachedImage.h:
* Source/WebCore/rendering/style/StyleImage.h:
(WebCore::StyleImage::isLoaded const):
(WebCore::StyleImage::reresolvedURL const):
* Source/WebCore/rendering/style/StyleMultiImage.cpp:
(WebCore::StyleMultiImage::isLoaded const):
* Source/WebCore/rendering/style/StyleMultiImage.h:
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::applyResource):
(WebCore::RenderSVGResourceMasker::drawContentIntoMaskImage):
(WebCore::RenderSVGResourceMasker::drawContentIntoContext):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.h:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeResources):
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::buildCachedResources):
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::rootElement const):
* Source/WebCore/svg/graphics/SVGImage.h:
* Source/WebCore/svg/graphics/SVGResourceImage.cpp: Added.
(WebCore::SVGResourceImage::create):
(WebCore::SVGResourceImage::SVGResourceImage):
(WebCore::SVGResourceImage::draw):
(WebCore::SVGResourceImage::drawPattern):
(WebCore::SVGResourceImage::dump const):
* Source/WebCore/svg/graphics/SVGResourceImage.h: Added.

Canonical link: <a href="https://commits.webkit.org/268272@main">https://commits.webkit.org/268272@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ebe56bf8a4458626234a7f84e8a8354e8d7af9b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19181 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20191 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21070 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17953 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19396 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22866 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19722 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19402 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19476 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16687 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21946 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16673 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17473 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23834 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17724 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/17647 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21788 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18234 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17357 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/17310 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4588 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21717 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18070 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/css/CSSImageValue.cpp:
```diff
@@ -85,16 +87,25 @@ URL CSSImageValue::reresolvedURL(const Document& document) const
+{
+    auto cssURLValue = location.resolvedURL.string();
+    auto fragmentIdentifier = SVGURIReference::fragmentIdentifierFromIRIString(cssURLValue, state.document());
+    if (fragmentIdentifier.isEmpty())
```
What happens if the URL is `foo.png#blah`? We should be ready for this: https://developer.mozilla.org/en-US/docs/Web/CSS/image/image#image_fragments

shallawa replied:
I added a test case and marked it [ ImageOnlyFailure ] and filed https://bugs.webkit.org/show_bug.cgi?id=261667 to track this issue.

shallawa replied:
Currently <img srd="foo.png#xywh=.."> shows the whole image. So mask-image: url(foo.png#blah); uses the whole image in masking the target element. Once the image fragment is supported, fixing the masking for this case should be straightforward.

smfr commented on Source/WebCore/rendering/ReferencedSVGResources.h:
```diff
@@ -59,6 +60,8 @@ class ReferencedSVGResources {
     RenderSVGResourceClipper* referencedClipperRenderer(TreeScope&, const ReferencePathOperation&);
     SVGFilterElement* referencedFilterElement(TreeScope&, const ReferenceFilterOperation&);
 
+    RenderSVGResourceContainer* referencedRenderResource(TreeScope&, const AtomString& fragment);
```
I think this could be `static` or at lest `const` ?

shallawa replied:
Fixed. It is made static.

smfr commented on Source/WebCore/rendering/style/StyleSVGResourceImage.h:
```diff
@@ -0,0 +1,67 @@
+    void didRemoveClient(RenderElement&) final { }
+
+    Ref<CSSImageValue> m_cssValue;
+    AtomString m_fragmentIdentifier;
```
I think we need to store the whole URL here to prepare for external resources, and to ensure that `url(foo.svg#maskThing)` compares differently to `url(#maskThing)`

shallawa replied:
This class is removed. The current patch handles both inline and external SVG masker resources.

smfr commented on Source/WebCore/rendering/style/StyleSVGResourceImage.h:
```diff
@@ -0,0 +1,67 @@
+    bool equals(const StyleSVGResourceImage&) const;
+    bool equalInputImages(const StyleSVGResourceImage&) const;
+
+    static constexpr bool isFixedSize = true;
```
Is it fixed size though?

shallawa replied:
This class is removed. And SVG maskers should not be fixed size.

smfr commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -107,15 +107,37 @@ bool RenderSVGResourceMasker::applyResource(RenderElement& renderer, const Rende
 {
-    GraphicsContext& maskImageContext = maskerData->maskImage->context();
+    auto& maskImageContext = maskerData->maskImage->context();
+    auto objectBoundingBox = object->objectBoundingBox();
```
This needs to respect `mask-origin`: https://developer.mozilla.org/en-US/docs/Web/CSS/mask-origin

I think you should share some code here with `BackgroundPainter::calculateBackgroundImageGeometry`

shallawa replied:
This can be handled by a separate patch. I filed https://bugs.webkit.org/show_bug.cgi?id=261670 to track this issue.

smfr commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -107,15 +107,37 @@ bool RenderSVGResourceMasker::applyResource(RenderElement& renderer, const Rende
+#endif
+
+    // Create the luminance mask.
+    if (style().svgStyle().maskType() == MaskType::Luminance)
```
Is this the right place to do this, if there's a list of mask images (only some of which might be SVG references)?

smfr commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -107,15 +107,37 @@ bool RenderSVGResourceMasker::applyResource(RenderElement& renderer, const Rende
         maskContentTransformation.scale(objectBoundingBox.size());
-        maskImageContext.concatCTM(maskContentTransformation);
+        context.concatCTM(maskContentTransformation);
+    } else if (effectiveZoom != 1) {
```
Are there tests with zooming?

shallawa replied:
I removed this part since this function is called with effectiveZoom = 1. Zooming is handled when drawing the SVG resource by scaling the destination rectangle.

smfr commented on Source/WebCore/svg/graphics/SVGResourceImage.h:
```diff
@@ -0,0 +1,55 @@
+    bool isSVGResourceImage() const final { return true; }
+    void dump(WTF::TextStream&) const final;
+
+    WeakPtr<RenderSVGResourceContainer> m_renderResource;
```
Seeing this weak ref makes me think we should have a test that dynamically sets the `<svg>` to `display:none` when some CSS references a mask inside of it.

smfr commented on LayoutTests/css3/masking/mask-reference-png-fragment.html:
```diff
@@ -0,0 +1,31 @@
+        background-color: blue;
+    }
+    .mask-bitmap-top-fragment {
+        mask-image: url(resources/red-100x100.png#xywh=0,0,100,50);
```
We don't support this fragment syntax, right?

shallawa replied:
Yes we do not support it right now. So I added this test. marked it as `[ ImageOnlyFailure ]` and filed [261667](https://bugs.webkit.org/show_bug.cgi?id=261667) to track this issue till it is fixed.

smfr commented on Source/WebCore/rendering/ReferencedSVGResources.h:
```diff
@@ -56,8 +56,10 @@ class ReferencedSVGResources {
-    RenderSVGResourceClipper* referencedClipperRenderer(TreeScope&, const ReferencePathOperation&);
-    SVGFilterElement* referencedFilterElement(TreeScope&, const ReferenceFilterOperation&);
+    static RenderSVGResourceClipper* referencedClipperRenderer(TreeScope&, const ReferencePathOperation&);
+    static SVGFilterElement* referencedFilterElement(TreeScope&, const ReferenceFilterOperation&);
```
This should return a `RefPtr<>`

smfr commented on Source/WebCore/rendering/style/StyleCachedImage.cpp:
```diff
@@ -68,7 +73,7 @@ bool StyleCachedImage::equals(const StyleCachedImage& other) const
     if (m_cssValue.ptr() == other.m_cssValue.ptr() || m_cssValue->equals(other.m_cssValue.get()))
         return true;
-    if (m_cachedImage && m_cachedImage == other.m_cachedImage)
+    if (cachedImage() == other.cachedImage())
```
Do you want both nulls to compare as equal? That's a behavior change.

shallawa replied:
I will revert this change.

smfr commented on Source/WebCore/rendering/style/StyleCachedImage.cpp:
```diff
@@ -149,10 +183,19 @@ bool StyleCachedImage::imageHasRelativeHeight() const
-void StyleCachedImage::computeIntrinsicDimensions(const RenderElement*, Length& intrinsicWidth, Length& intrinsicHeight, FloatSize& intrinsicRatio)
+void StyleCachedImage::computeIntrinsicDimensions(const RenderElement* renderer, Length& intrinsicWidth, Length& intrinsicHeight, FloatSize& intrinsicRatio)
 {
+    if (renderSVGResource(renderer)) {
```
I'd appreciate a comment saying what this size is computed as.

nt1m commented on LayoutTests/css3/masking/mask-reference-png-fragment-expected.html:
```diff
@@ -0,0 +1,32 @@
+        border: 1px solid black;
+    }
+    .box {
+        position: rekative;
```
```suggestion
        position: relative;
```

shallawa committed:
[CSS Masking] SVG masks are not working as 'mask-image'
https://bugs.webkit.org/show_bug.cgi?id=260732
rdar://114465545

Reviewed by Simon Fraser.

Implement CSS masking when it references an SVG masker. There seven cases for
mask-image values that need to be considered:

1. Bitmap image
2. Bitmap image fragment
3. External SVG image
4. External SVG image fragment
5. External SVG image mask
6. Inline data URL image
7. Inline SVG image mask

StyleCachedImage will be adjusted to take into account the seventh case. Functions
like StyleCachedImage::isLoaded() and StyleCachedImage::canRender() will have to
handle the new case where StyleCachedImage represents an inline referenced SVG
masker. So isLoaded() needs to a take a RenderElement to be able to resolve the
imageURL.

A new super-class of GeneratedImage named SVGResourceImage will be introduced to
handle drawing the SVG resource given its renderer. SVGResourceImage will handle
the cases of external and inline SVG resources.

We can't know whether the ID associated with an external image is a fragment ID or
a mask ID in an SVGImage till the CachedImage is loaded. StyleCachedImage::image()
is called only after the image is completely loaded. So it can decide to create
an SVGResourceImage or to let the CachedImage creates its Image.

* LayoutTests/TestExpectations:
Some of the WPT tests have been marked as ImageOnlyFailure in this patch because
the expected results are now working as expected but the tests are not.
Unfortunately the tests and the expected results are both using mask-image. So
without this change they match because they both draw nothing for the masked
elements. With this change, the tests are failing  because the mask geometry
is set to work with content-box only. This can be addressed in a separate patch.

* LayoutTests/css3/masking/mask-reference-png-fragment-expected.html: Added.
* LayoutTests/css3/masking/mask-reference-png-fragment.html: Added.
* LayoutTests/css3/masking/resources/red-100x100.png: Added.
* LayoutTests/http/tests/css/css-masking/mask-external-svg-fragment-expected.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-external-svg-fragment.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-external-svg-image-expected.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-external-svg-image.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-external-svg-mask-expected.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-external-svg-mask.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-inline-svg-image-expected.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-inline-svg-image.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-inline-svg-mask-expected.html: Added.
* LayoutTests/http/tests/css/css-masking/mask-inline-svg-mask.html: Added.
* LayoutTests/http/tests/css/css-masking/resources/heart.svg: Added.
* LayoutTests/http/tests/css/css-masking/resources/star-heart-fragments.svg: Added.
* LayoutTests/http/tests/css/css-masking/resources/star-heart-masks.svg: Added.
* LayoutTests/http/tests/css/css-masking/resources/star.svg: Added.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/mask-image/mask-opacity-1d.html:
This tests is now drawing the masked element but with noise around the borders.

* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSImageValue.cpp:
* Source/WebCore/platform/graphics/Image.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::isSVGResourceImage const):
* Source/WebCore/rendering/BorderPainter.cpp:
(WebCore::BorderPainter::paintBorder):
(WebCore::BorderPainter::paintNinePieceImage):
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::referenceFilterElement):
* Source/WebCore/rendering/InlineBoxPainter.cpp:
(WebCore::InlineBoxPainter::paintMask):
(WebCore::InlineBoxPainter::paintDecorations):
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
(WebCore::ReferencedSVGResources::referencedRenderResource):
* Source/WebCore/rendering/ReferencedSVGResources.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::paintMaskImages):
(WebCore::RenderBox::repaintLayerRectsForImage):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::borderImageIsLoadedAndCanBeRendered const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::setupClipPath):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::canDirectlyCompositeBackgroundBackgroundImage):
(WebCore::hasPaintedBoxDecorationsOrBackgroundImage):
(WebCore::supportsDirectlyCompositedBoxDecorations):
(WebCore::RenderLayerBacking::contentChanged):
* Source/WebCore/rendering/style/FillLayer.cpp:
(WebCore::FillLayer::imagesAreLoaded const):
* Source/WebCore/rendering/style/FillLayer.h:
* Source/WebCore/rendering/style/NinePieceImage.cpp:
(WebCore::NinePieceImage::paint const):
* Source/WebCore/rendering/style/StyleCachedImage.cpp:
(WebCore::StyleCachedImage::renderSVGResource const):
(WebCore::StyleCachedImage::canRender const):
(WebCore::StyleCachedImage::isLoaded const):
(WebCore::StyleCachedImage::imageSize const):
(WebCore::StyleCachedImage::computeIntrinsicDimensions):
(WebCore::StyleCachedImage::setContainerContextForRenderer):
(WebCore::StyleCachedImage::image const):
* Source/WebCore/rendering/style/StyleCachedImage.h:
* Source/WebCore/rendering/style/StyleImage.h:
(WebCore::StyleImage::isLoaded const):
(WebCore::StyleImage::reresolvedURL const):
* Source/WebCore/rendering/style/StyleMultiImage.cpp:
(WebCore::StyleMultiImage::isLoaded const):
* Source/WebCore/rendering/style/StyleMultiImage.h:
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::applyResource):
(WebCore::RenderSVGResourceMasker::drawContentIntoMaskImage):
(WebCore::RenderSVGResourceMasker::drawContentIntoContext):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.h:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeResources):
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::buildCachedResources):
* Source/WebCore/svg/graphics/SVGImage.cpp:
(WebCore::SVGImage::rootElement const):
* Source/WebCore/svg/graphics/SVGImage.h:
* Source/WebCore/svg/graphics/SVGResourceImage.cpp: Added.
(WebCore::SVGResourceImage::create):
(WebCore::SVGResourceImage::SVGResourceImage):
(WebCore::SVGResourceImage::draw):
(WebCore::SVGResourceImage::drawPattern):
(WebCore::SVGResourceImage::dump const):
* Source/WebCore/svg/graphics/SVGResourceImage.h: Added.

Canonical link: https://commits.webkit.org/268272@main
